shared/logger.go

```diff
@@ -0,0 +1,219 @@
+
+// LoggerCtxKey is a key for attaching a Logger to a context.Context.
+type LoggerCtxKey struct{}
+
+var (
+	gl  = gaeLogger{}
+	nl  = nilLogger{}
+	lck = LoggerCtxKey{}
+)
+
+// NewLoggerMux creates a multiplexing Logger that writes all log operations to
+// all delegates.
+func NewLoggerMux(delegates []Logger) Logger {
+	if len(delegates) == 0 {
+		return NewNilLogger()
+	}
+	return loggerMux{delegates}
+}
+
+// NewGAELogger returns a Google App Engine Standard Environment logger bound to
+// the given context.
+func NewGAELogger(ctx context.Context) Logger {
+	return gaeLogger{ctx}
+}
+
+// NewNilLogger returns a new logger that silently ignores all Logger calls.
+func NewNilLogger() Logger {
+	return nl
+}
+
+// DefaultLoggerCtxKey returns the default key where a logger instance should be
+// stored in a context.Context object.
+func DefaultLoggerCtxKey() LoggerCtxKey {
+	return lck
+}
+
+// WithLogger is a strongly-typed setter for the logger value on the context.
+func WithLogger(ctx context.Context, logger Logger) context.Context {
+	return context.WithValue(ctx, DefaultLoggerCtxKey(), logger)
+}
+
+// GetLogger retrieves a non-nil Logger that is appropriate for use in ctx. If
+// ctx does not provide a logger, then a nil-logger is returned.
+func GetLogger(ctx context.Context) Logger {
+	logger, ok := ctx.Value(DefaultLoggerCtxKey()).(Logger)
+	if !ok || logger == nil {
+		log.Warningf("Context without logger: %v; logs will be dropped", ctx)
+		return NewNilLogger()
+	}
+
+	return logger
+}
+
+type cloudLoggingClientKeyType struct{}
+
+var cloudLoggingClientKey = cloudLoggingClientKeyType{}
+
+type traceKeyType struct{}
+
+var traceKey = traceKeyType{}
+
+// HandleWithGoogleCloudLogging handles the request with the given handler, setting the logger
+// on the request's context to be a Google Cloud logging client for the given project.
+//
+// commonResource is an optional override to the monitored resource details appended to each log.
+// e.g. in the Flex environment, it pays to override this value to type gae_app, to ensure finding
+// logs is consistent between services.
+func HandleWithGoogleCloudLogging(h http.HandlerFunc, project string, commonResource *mrpb.MonitoredResource) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx, err := NewAppEngineFlexContext(r, project, commonResource)
+		if err != nil {
+			h(w, r)
+			return
+		}
+		h(w, r.WithContext(ctx))
+	}
+}
+
+// NewAppEngineFlexContext creates a new Google App Engine Flex-based
+// context, with a Google Cloud logger client bound to an http.Request.
+func NewAppEngineFlexContext(r *http.Request, project string, commonResource *mrpb.MonitoredResource) (ctx context.Context, err error) {
+	ctx = r.Context()
+	client, err := gclog.NewClient(ctx, project)
+	if err != nil {
+		return nil, err
+	}
+	ctx = context.WithValue(ctx, cloudLoggingClientKey, client)
+
+	traceID := strings.Split(r.Header.Get("X-Cloud-Trace-Context"), "/")[0]
```

Comment:
Nit: perhaps add a comment and link to https://cloud.google.com/appengine/docs/flexible/go/writing-application-logs